File: frontend/src/lib/queries/hosted-zones.ts
'use client';

import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';

import {
  createHostedZone,
  deleteHostedZone,
  getHostedZone,
  listHostedZones,
  updateHostedZone,
} from '@/lib/api';
import type {
  HostedZoneCreate,
  HostedZoneResponse,
  ZoneSortField,
  SortOrder,
} from '@/lib/api/types.gen';
import { toUserMessage } from '@/lib/queries/client';
import { queryKeys } from '@/lib/queries/keys';
import { notify } from '@/lib/notifications';

export interface ZoneListParams {
  search?: string;
  sort: ZoneSortField;
  order: SortOrder;
  limit: number;
  offset: number;
}

/** Fetches one page of hosted zones. */
export function useHostedZones(params: ZoneListParams) {
  return useQuery({
    queryKey: queryKeys.zones.list(params),
    queryFn: async () => {
      const { data, error } = await listHostedZones({ query: params });
      if (error) throw error;
      return data;
    },
    // Keeps the previous page visible while the next loads, so paging does not
    // flash an empty table.
    placeholderData: (previous) => previous,
  });
}

/** Fetches one hosted zone by its public identifier. */
export function useHostedZone(zoneId: string) {
  return useQuery({
    queryKey: queryKeys.zones.detail(zoneId),
    queryFn: async () => {
      const { data, error } = await getHostedZone({ path: { zone_id: zoneId } });
      if (error) throw error;
      return data;
    },
    enabled: Boolean(zoneId),
  });
}

/** Creates a hosted zone, then invalidates every zone listing. */
export function useCreateHostedZone() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (body: HostedZoneCreate) => {
      const { data, error } = await createHostedZone({ body });
      if (error) throw error;
      return data;
    },
    onSuccess: (zone: HostedZoneResponse) => {
      void queryClient.invalidateQueries({ queryKey: queryKeys.zones.all() });
      notify(
        'success',
        `Hosted zone ${zone.name} created`,
        'An SOA record and an NS record listing four name servers were created with it.',
      );
    },
    onError: (error) => notify('error', 'Could not create the hosted zone', toUserMessage(error)),
  });
}

/** Updates a hosted zone's description. */
export function useUpdateHostedZone(zoneId: string) {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (comment: string | null) => {
      const { data, error } = await updateHostedZone({
        path: { zone_id: zoneId },
        body: { comment },
      });
      if (error) throw error;
      return data;
    },
    onSuccess: () => {
      void queryClient.invalidateQueries({ queryKey: queryKeys.zones.all() });
      notify('success', 'Description updated');
    },
    onError: (error) => notify('error', 'Could not update the hosted zone', toUserMessage(error)),
  });
}

/** Deletes a hosted zone. */
export function useDeleteHostedZone() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (zone: HostedZoneResponse) => {
      const { error } = await deleteHostedZone({ path: { zone_id: zone.id } });
      if (error) throw error;
      return zone;
    },
    onSuccess: (zone) => {
      void queryClient.invalidateQueries({ queryKey: queryKeys.zones.all() });
      notify('success', `Hosted zone ${zone.name} deleted`);
    },
    onError: (error) => notify('error', 'Could not delete the hosted zone', toUserMessage(error)),
  });
}

File: backend/app/api/v1/hosted_zones.py
"""Hosted-zone endpoints.

Routers stay thin: parse, delegate, return. Business rules live in the service
and error translation lives in the exception handlers, so there is no try/except
anywhere in this file.
"""

from typing import Annotated

from fastapi import APIRouter, Depends, Query, status

from app.api.v1.dependencies import CurrentUser, HostedZoneServiceDep, require_csrf_token
from app.repositories.protocols import SortOrder, ZoneSortField
from app.schemas.common import ERROR_RESPONSES, MAX_PAGE_SIZE, Page, build_page
from app.schemas.hosted_zone import HostedZoneCreate, HostedZoneResponse, HostedZoneUpdate

router = APIRouter(prefix="/hosted-zones", tags=["Hosted zones"])


@router.get(
    "",
    response_model=Page[HostedZoneResponse],
    summary="List hosted zones",
    description=(
        "Returns the signed-in user's hosted zones. Sorting is restricted to an "
        "allowlist of columns, so no client value ever reaches the ORDER BY clause "
        "unvalidated."
    ),
    responses={401: ERROR_RESPONSES[401]},
)
def list_hosted_zones(
    user: CurrentUser,
    service: HostedZoneServiceDep,
    search: Annotated[str | None, Query(max_length=255, description="Substring of the name.")] = None,
    sort: ZoneSortField = ZoneSortField.NAME,
    order: SortOrder = SortOrder.ASC,
    limit: Annotated[int, Query(ge=1, le=MAX_PAGE_SIZE)] = 10,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> Page[HostedZoneResponse]:
    """Return one page of the user's hosted zones."""
    zones, total = service.list_zones(
        user.id, search=search, sort=sort, order=order, limit=limit, offset=offset
    )
    items = [HostedZoneResponse.model_validate(zone) for zone in zones]
    return build_page(items, total, limit, offset)


@router.post(
    "",
    response_model=HostedZoneResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create a hosted zone",
    description=(
        "Creates a zone together with its generated SOA record and apex NS record "
        "listing four name servers, in one transaction. Both are read-only "
        "afterwards."
    ),
    responses={
        401: ERROR_RESPONSES[401],
        409: {**ERROR_RESPONSES[409], "description": "You already have a zone with this name."},
        422: ERROR_RESPONSES[422],
    },
    dependencies=[Depends(require_csrf_token)],
)
def create_hosted_zone(
    payload: HostedZoneCreate,
    user: CurrentUser,
    service: HostedZoneServiceDep,
) -> HostedZoneResponse:
    """Create a hosted zone and its system records."""
    zone = service.create_zone(
        user.id, name=payload.name, comment=payload.comment, zone_type=payload.type
    )
    return HostedZoneResponse.model_validate(zone)


@router.get(
    "/{zone_id}",
    response_model=HostedZoneResponse,
    summary="Get a hosted zone",
    description=(
        "Returns one zone. A zone owned by another user produces 404 rather than "
        "403, because 403 would confirm that the identifier exists."
    ),
    responses={401: ERROR_RESPONSES[401], 404: ERROR_RESPONSES[404]},
)
def get_hosted_zone(
    zone_id: str,
    user: CurrentUser,
    service: HostedZoneServiceDep,
) -> HostedZoneResponse:
    """Return one hosted zone."""
    return HostedZoneResponse.model_validate(service.get_zone(zone_id, user.id))


@router.patch(
    "/{zone_id}",
    response_model=HostedZoneResponse,
    summary="Update a hosted zone",
    description="Updates the description. A zone's name is immutable.",
    responses={401: ERROR_RESPONSES[401], 404: ERROR_RESPONSES[404], 422: ERROR_RESPONSES[422]},
    dependencies=[Depends(require_csrf_token)],
)
def update_hosted_zone(
    zone_id: str,
    payload: HostedZoneUpdate,
    user: CurrentUser,
    service: HostedZoneServiceDep,
) -> HostedZoneResponse:
    """Update a hosted zone's description."""
    zone = service.update_zone(zone_id, user.id, comment=payload.comment)
    return HostedZoneResponse.model_validate(zone)


@router.delete(
    "/{zone_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete a hosted zone",
    description=(
        "Deletes a zone that holds no records beyond its generated SOA and NS. "
        "Delete the remaining records first."
    ),
    responses={
        401: ERROR_RESPONSES[401],
        404: ERROR_RESPONSES[404],
        409: {**ERROR_RESPONSES[409], "description": "The zone still contains records."},
    },
    dependencies=[Depends(require_csrf_token)],
)
def delete_hosted_zone(
    zone_id: str,
    user: CurrentUser,
    service: HostedZoneServiceDep,
) -> None:
    """Delete an empty hosted zone."""
    service.delete_zone(zone_id, user.id)
